// materialize-angular/esm5/app/completed-components/button/button.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/button/button.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
/**
 * @record
 */
export function ButtonModel() { }
if (false) {
    /** @type {?} */
    ButtonModel.prototype.className;
    /** @type {?} */
    ButtonModel.prototype.disabled;
    /** @type {?} */
    ButtonModel.prototype.level;
    /** @type {?} */
    ButtonModel.prototype.rippleDuration;
    /** @type {?} */
    ButtonModel.prototype.type;
}
/** @enum {string} */
var BUTTON_LEVELS = {
    TEXT: "text",
    ACCENT: "accent",
    OUTLINE: "outline",
};
export { BUTTON_LEVELS };
/** @enum {string} */
var BUTTON_TYPES = {
    BUTTON: "button",
    SUBMIT: "submit",
    RESET: "reset",
};
export { BUTTON_TYPES };
