// materialize-angular/esm5/app/completed-components/scroll-spy/scroll-spy.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/scroll-spy/scroll-spy.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ContentChildren, EventEmitter, Input, Output, QueryList } from '@angular/core';
import { config } from '../../config';
import { getOffseTop } from '../../utils/get-offset-top.util';
import { ScrollSpyItemComponent } from './scroll-spy-item/scroll-spy-item.component';
import { windowScrollPosition } from '../../utils/window-scroll-position.util';
var ScrollSpyComponent = /** @class */ (function () {
    function ScrollSpyComponent() {
        this.className = ScrollSpyComponent.defaultProps.className;
        this.itemId = ScrollSpyComponent.defaultProps.itemId;
        this.onChangeEmitter = new EventEmitter();
        this.onScroll = this.onScroll.bind(this);
        this.validateScroll = this.validateScroll.bind(this);
        window.addEventListener('scroll', this.onScroll);
    }
    /**
     * @return {?}
     */
    ScrollSpyComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.onScroll();
    };
    /**
     * @return {?}
     */
    ScrollSpyComponent.prototype.onScroll = /**
     * @return {?}
     */
    function () {
        setTimeout(this.validateScroll, 100);
    };
    /**
     * @return {?}
     */
    ScrollSpyComponent.prototype.validateScroll = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var item = this.getCurrentItem();
        if (this.itemId !== item.id) {
            this.itemId = item.id;
            this.onChangeEmitter.emit(this.itemId);
        }
    };
    /**
     * @return {?}
     */
    ScrollSpyComponent.prototype.getCurrentItem = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var items = this.items.toArray();
        for (var i = 0; i < items.length; i++) {
            /** @type {?} */
            var elementOffsetTop = getOffseTop(items[i].element.nativeElement);
            if (elementOffsetTop > windowScrollPosition()) {
                /** @type {?} */
                var prevItem = items[i - 1]
                    ? items[i - 1]
                    : items[0];
                return prevItem;
            }
        }
        return items[items.length - 1];
    };
    ScrollSpyComponent.defaultProps = {
        className: '',
        itemId: ''
    };
    ScrollSpyComponent.decorators = [
        { type: Component, args: [{
                    selector: config.components.prefix + "-scroll-spy",
                    template: "<div [ngClass]=\"className\">\n  <ng-content select=\"materialize-scroll-spy-item\"></ng-content>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    ScrollSpyComponent.ctorParameters = function () { return []; };
    ScrollSpyComponent.propDecorators = {
        items: [{ type: ContentChildren, args: [ScrollSpyItemComponent,] }],
        onChangeEmitter: [{ type: Output, args: ['onChange',] }],
        className: [{ type: Input }],
        itemId: [{ type: Input }]
    };
    return ScrollSpyComponent;
}());
export { ScrollSpyComponent };
if (false) {
    /** @type {?} */
    ScrollSpyComponent.defaultProps;
    /** @type {?} */
    ScrollSpyComponent.prototype.items;
    /** @type {?} */
    ScrollSpyComponent.prototype.onChangeEmitter;
    /** @type {?} */
    ScrollSpyComponent.prototype.className;
    /** @type {?} */
    ScrollSpyComponent.prototype.itemId;
}
