// materialize-angular/esm5/app/completed-components/stepper/step/step.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/stepper/step/step.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, TemplateRef, ViewChild } from '@angular/core';
import { config } from '../../../config';
var StepComponent = /** @class */ (function () {
    function StepComponent() {
        this.prefix = config.components.prefix;
        this.isCompleted = false;
    }
    StepComponent.decorators = [
        { type: Component, args: [{
                    selector: config.components.prefix + "-step }",
                    template: "<ng-template #indicatorTemplate>\n  <div [ngClass]=\"prefix + '-step-indicator'\">\n    <materialize-icon [className]=\"prefix + '-step-indicator-icon'\" *ngIf=\"isCompleted\">done</materialize-icon>\n    <ng-content select=\"materialize-step-indicator\" *ngIf=\"!isCompleted\"></ng-content>\n  </div>\n</ng-template>\n<ng-template #labelTemplate>\n  <ng-content select=\"materialize-step-label\"></ng-content>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    StepComponent.ctorParameters = function () { return []; };
    StepComponent.propDecorators = {
        labelTemplate: [{ type: ViewChild, args: ['labelTemplate', { static: false },] }],
        indicatorTemplate: [{ type: ViewChild, args: ['indicatorTemplate', { static: false },] }]
    };
    return StepComponent;
}());
export { StepComponent };
if (false) {
    /** @type {?} */
    StepComponent.prototype.labelTemplate;
    /** @type {?} */
    StepComponent.prototype.indicatorTemplate;
    /** @type {?} */
    StepComponent.prototype.prefix;
    /** @type {?} */
    StepComponent.prototype.isCompleted;
}
