// materialize-angular/esm5/app/completed-components/select/select-option/select-option.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/select/select-option/select-option.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { config } from '../../../config';
var SelectOptionComponent = /** @class */ (function () {
    function SelectOptionComponent() {
        this.className = SelectOptionComponent.defaultProps.className;
        this.disabled = SelectOptionComponent.defaultProps.disabled;
        this.value = SelectOptionComponent.defaultProps.value;
        this.prefix = config.components.prefix;
        this.transitionDuration = SelectOptionComponent.defaultProps.transitionDuration;
        this.isActive = false;
        this.onClickEmitter = new EventEmitter();
        this.emitClick = this.emitClick.bind(this);
    }
    /**
     * @return {?}
     */
    SelectOptionComponent.prototype.ngAfterContentChecked = /**
     * @return {?}
     */
    function () {
        this.content = this.optionTemplateRef.nativeElement
            ? this.optionTemplateRef.nativeElement.textContent
            : '';
    };
    /**
     * @return {?}
     */
    SelectOptionComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (!this.disabled) {
            setTimeout(this.emitClick, this.transitionDuration);
        }
    };
    /**
     * @return {?}
     */
    SelectOptionComponent.prototype.emitClick = /**
     * @return {?}
     */
    function () {
        this.onClickEmitter.emit(this.value);
    };
    SelectOptionComponent.defaultProps = {
        className: '',
        disabled: false,
        transitionDuration: 250,
        value: ''
    };
    SelectOptionComponent.decorators = [
        { type: Component, args: [{
                    selector: config.components.prefix + "-select-option }",
                    template: "<div [ngClass]=\"[prefix + '-select-option', className]\" materializeRipple (click)=\"onClick()\" [class.active]=\"isActive\"  [class.disabled]=\"disabled\" [rippleDuration]=\"transitionDuration\">\n  <div #optionTemplate>\n    <span>\n      <ng-content></ng-content>\n    </span>\n  </div>\n</div>\n\n"
                }] }
    ];
    /** @nocollapse */
    SelectOptionComponent.ctorParameters = function () { return []; };
    SelectOptionComponent.propDecorators = {
        optionTemplateRef: [{ type: ViewChild, args: ['optionTemplate', { static: true },] }],
        onClickEmitter: [{ type: Output, args: ['onClick',] }],
        className: [{ type: Input }],
        disabled: [{ type: Input }],
        value: [{ type: Input }]
    };
    return SelectOptionComponent;
}());
export { SelectOptionComponent };
if (false) {
    /** @type {?} */
    SelectOptionComponent.defaultProps;
    /** @type {?} */
    SelectOptionComponent.prototype.optionTemplateRef;
    /** @type {?} */
    SelectOptionComponent.prototype.onClickEmitter;
    /** @type {?} */
    SelectOptionComponent.prototype.className;
    /** @type {?} */
    SelectOptionComponent.prototype.disabled;
    /** @type {?} */
    SelectOptionComponent.prototype.value;
    /** @type {?} */
    SelectOptionComponent.prototype.prefix;
    /** @type {?} */
    SelectOptionComponent.prototype.content;
    /** @type {?} */
    SelectOptionComponent.prototype.isActive;
    /** @type {?} */
    SelectOptionComponent.prototype.transitionDuration;
}
