// materialize-angular/esm5/app/completed-components/accordion/accordion.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/accordion/accordion.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ContentChildren, EventEmitter, Input, Output, QueryList } from '@angular/core';
import { CollapsibleComponent } from '../collapsible/collapsible.component';
import { config } from '../../config';
var AccordionComponent = /** @class */ (function () {
    function AccordionComponent() {
        this.className = AccordionComponent.defaultProps.className;
        this.activeIndex = AccordionComponent.defaultProps.activeIndex;
        this.initCollapsibles = this.initCollapsibles.bind(this);
        this.registerCollapsibles = this.registerCollapsibles.bind(this);
        this.onToggleEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AccordionComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.initCollapsibles();
        this.collapsiblesQueryList.changes.subscribe(this.initCollapsibles);
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.initCollapsibles = /**
     * @return {?}
     */
    function () {
        setTimeout(this.registerCollapsibles, 0);
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.registerCollapsibles = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.collapsiblesQueryList.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            item.onOpenEmitter.subscribe((/**
             * @return {?}
             */
            function () {
                _this.toggleCollapsibles(index);
            }));
        }));
    };
    /**
     * @param {?} currentIndex
     * @return {?}
     */
    AccordionComponent.prototype.toggleCollapsibles = /**
     * @param {?} currentIndex
     * @return {?}
     */
    function (currentIndex) {
        this.activeIndex = currentIndex;
        this.onToggleEmitter.emit(this.activeIndex);
        this.collapsiblesQueryList.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            if (index === currentIndex) {
                item.open();
            }
            else {
                item.close();
            }
        }));
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.showNext = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var nextIndex = this.activeIndex === null || this.activeIndex >= this.collapsiblesQueryList.length
            ? null
            : this.activeIndex + 1;
        this.toggleCollapsibles(nextIndex);
    };
    /**
     * @return {?}
     */
    AccordionComponent.prototype.showPrev = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var prevIndex = this.activeIndex === null || this.activeIndex === 0
            ? null
            : this.activeIndex - 1;
        this.toggleCollapsibles(prevIndex);
    };
    AccordionComponent.defaultProps = {
        activeIndex: null,
        className: ''
    };
    AccordionComponent.decorators = [
        { type: Component, args: [{
                    selector: config.components.prefix + "-accordion }",
                    template: "<div [ngClass]=\"className\">\n  <ng-content select=\"materialize-collapsible\"></ng-content>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    AccordionComponent.ctorParameters = function () { return []; };
    AccordionComponent.propDecorators = {
        collapsiblesQueryList: [{ type: ContentChildren, args: [CollapsibleComponent,] }],
        onToggleEmitter: [{ type: Output, args: ['onToggle',] }],
        className: [{ type: Input }]
    };
    return AccordionComponent;
}());
export { AccordionComponent };
if (false) {
    /** @type {?} */
    AccordionComponent.defaultProps;
    /** @type {?} */
    AccordionComponent.prototype.collapsiblesQueryList;
    /** @type {?} */
    AccordionComponent.prototype.onToggleEmitter;
    /** @type {?} */
    AccordionComponent.prototype.className;
    /** @type {?} */
    AccordionComponent.prototype.activeIndex;
}
