// materialize-angular/esm5/app/completed-components/modal/modal.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/modal/modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { config } from '../../config';
var ModalComponent = /** @class */ (function () {
    function ModalComponent() {
        this.className = ModalComponent.defaultProps.className;
        this.dismissOnBackdrop = ModalComponent.defaultProps.dismissOnBackdrop;
        this.hasBackdrop = ModalComponent.defaultProps.hasBackdrop;
        this.hasCloseButton = ModalComponent.defaultProps.hasCloseButton;
        this.isOpen = ModalComponent.defaultProps.isOpen;
        this.transitionDuration = ModalComponent.defaultProps.transitionDuration;
        this.prefix = config.components.prefix;
        this.onOpenEmitter = new EventEmitter();
        this.onCloseEmitter = new EventEmitter();
        this.open = this.open.bind(this);
        this.close = this.close.bind(this);
        this.closeByBackdrop = this.closeByBackdrop.bind(this);
    }
    /**
     * @return {?}
     */
    ModalComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.hasBackdrop && this.dismissOnBackdrop) {
            this.backdropRef.nativeElement.addEventListener('click', this.closeByBackdrop);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ModalComponent.prototype.closeByBackdrop = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var target = event.target;
        var nativeElement = this.backdropRef.nativeElement;
        if (target === nativeElement) {
            this.close();
        }
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.modalRef.nativeElement.style.transitionDuration = this.transitionDuration + "ms";
        this.isOpen = true;
        this.onOpenEmitter.emit();
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.modalRef.nativeElement.style.transitionDuration = this.transitionDuration / 2 + "ms";
        this.isOpen = false;
        this.onCloseEmitter.emit();
    };
    ModalComponent.defaultProps = {
        className: '',
        dismissOnBackdrop: true,
        hasBackdrop: true,
        hasCloseButton: true,
        isOpen: false,
        transitionDuration: 400
    };
    ModalComponent.decorators = [
        { type: Component, args: [{
                    selector: config.components.prefix + "-modal }",
                    template: "<div [ngClass]=\"[prefix + '-modal', className]\" [class.active]=\"isOpen\" #modal>\n  <div [ngClass]=\"prefix + '-modal-backdrop'\" #backdrop *ngIf=\"hasBackdrop\"></div>\n  <div [ngClass]=\"prefix + '-modal-content'\">\n    <materialize-icon [className]=\"prefix + '-modal-icon-close'\" *ngIf=\"hasCloseButton\" (onClick)=\"close()\">close</materialize-icon>\n\n    <ng-content></ng-content>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    ModalComponent.ctorParameters = function () { return []; };
    ModalComponent.propDecorators = {
        modalRef: [{ type: ViewChild, args: ['modal', { static: true },] }],
        backdropRef: [{ type: ViewChild, args: ['backdrop', { static: false },] }],
        onOpenEmitter: [{ type: Output, args: ['onOpen',] }],
        onCloseEmitter: [{ type: Output, args: ['onClose',] }],
        className: [{ type: Input }],
        dismissOnBackdrop: [{ type: Input }],
        hasBackdrop: [{ type: Input }],
        hasCloseButton: [{ type: Input }],
        isOpen: [{ type: Input }],
        transitionDuration: [{ type: Input }]
    };
    return ModalComponent;
}());
export { ModalComponent };
if (false) {
    /** @type {?} */
    ModalComponent.defaultProps;
    /** @type {?} */
    ModalComponent.prototype.modalRef;
    /** @type {?} */
    ModalComponent.prototype.backdropRef;
    /** @type {?} */
    ModalComponent.prototype.onOpenEmitter;
    /** @type {?} */
    ModalComponent.prototype.onCloseEmitter;
    /** @type {?} */
    ModalComponent.prototype.className;
    /** @type {?} */
    ModalComponent.prototype.dismissOnBackdrop;
    /** @type {?} */
    ModalComponent.prototype.hasBackdrop;
    /** @type {?} */
    ModalComponent.prototype.hasCloseButton;
    /** @type {?} */
    ModalComponent.prototype.isOpen;
    /** @type {?} */
    ModalComponent.prototype.transitionDuration;
    /** @type {?} */
    ModalComponent.prototype.prefix;
}
