// materialize-angular/esm2015/app/completed-components/collapsible/collapsible.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/collapsible/collapsible.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
import { config } from '../../config';
export class CollapsibleComponent {
    /**
     * @param {?} renderer
     */
    constructor(renderer) {
        this.renderer = renderer;
        this.className = CollapsibleComponent.defaultProps.className;
        this.disabled = CollapsibleComponent.defaultProps.disabled;
        this.isOpen = CollapsibleComponent.defaultProps.isOpen;
        this.showIndicator = CollapsibleComponent.defaultProps.showIndicator;
        this.prefix = config.components.prefix;
        this.onCloseEmitter = new EventEmitter();
        this.onOpenEmitter = new EventEmitter();
        this.onToggle = this.onToggle.bind(this);
        this.update = this.update.bind(this);
        window.addEventListener('resize', this.update);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (this.isOpen) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.update();
            }), 300);
        }
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        this.update();
    }
    /**
     * @return {?}
     */
    onToggle() {
        if (!this.disabled) {
            if (this.isOpen) {
                this.close();
                this.onCloseEmitter.emit();
            }
            else {
                this.open();
                this.onOpenEmitter.emit();
            }
        }
    }
    /**
     * @return {?}
     */
    open() {
        this.isOpen = true;
        this.update();
    }
    /**
     * @return {?}
     */
    close() {
        this.isOpen = false;
        this.update();
    }
    /**
     * @return {?}
     */
    update() {
        /** @type {?} */
        const contentContainer = this.containerRef.nativeElement;
        /** @type {?} */
        const maxHeight = this.isOpen
            ? contentContainer.scrollHeight
            : 0;
        this.renderer.setStyle(contentContainer, 'maxHeight', `${maxHeight}px`);
    }
}
CollapsibleComponent.defaultProps = {
    className: '',
    disabled: false,
    isOpen: false,
    showIndicator: true
};
CollapsibleComponent.decorators = [
    { type: Component, args: [{
                selector: `${config.components.prefix}-collapsible }`,
                template: "<div\n  [ngClass]=\"[prefix + '-collapsible', className]\"\n  [class.disabled]=\"disabled\"\n  [class.open]=\"isOpen\"\n>\n  <div (click)=\"onToggle()\"\n    [isRippleActive]=\"!disabled\"\n    [ngClass]=\"[prefix + '-collapsible-title', className]\"\n\n    materializeRipple\n  >\n    <ng-content select=\"materialize-collapsible-title\"></ng-content>\n\n    <materialize-icon *ngIf=\"showIndicator\" [className]=\"prefix + '-collapsible-title-icon'\">\n      expand_more\n    </materialize-icon>\n  </div>\n\n  <div [ngClass]=\"[prefix + '-collapsible-content-container', className]\" #container>\n    <div [ngClass]=\"prefix + '-collapsible-content'\">\n      <ng-content select=\"materialize-collapsible-content\"></ng-content>\n    </div>\n  </div>\n</div>\n"
            }] }
];
/** @nocollapse */
CollapsibleComponent.ctorParameters = () => [
    { type: Renderer2 }
];
CollapsibleComponent.propDecorators = {
    onCloseEmitter: [{ type: Output, args: ['onClose',] }],
    onOpenEmitter: [{ type: Output, args: ['onOpen',] }],
    containerRef: [{ type: ViewChild, args: ['container', { static: true },] }],
    className: [{ type: Input }],
    disabled: [{ type: Input }],
    isOpen: [{ type: Input }],
    showIndicator: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CollapsibleComponent.defaultProps;
    /** @type {?} */
    CollapsibleComponent.prototype.onCloseEmitter;
    /** @type {?} */
    CollapsibleComponent.prototype.onOpenEmitter;
    /** @type {?} */
    CollapsibleComponent.prototype.containerRef;
    /** @type {?} */
    CollapsibleComponent.prototype.className;
    /** @type {?} */
    CollapsibleComponent.prototype.disabled;
    /** @type {?} */
    CollapsibleComponent.prototype.isOpen;
    /** @type {?} */
    CollapsibleComponent.prototype.showIndicator;
    /** @type {?} */
    CollapsibleComponent.prototype.prefix;
    /**
     * @type {?}
     * @private
     */
    CollapsibleComponent.prototype.renderer;
}
