// materialize-angular/esm2015/app/completed-components/form-message/form-message.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/form-message/form-message.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, Input } from '@angular/core';
import { FORM_MESSAGE_LEVELS } from './form-message.model';
import { config } from '../../config';
export class FormMessageComponent {
    constructor() {
        this.className = FormMessageComponent.defaultProps.className;
        this.level = FormMessageComponent.defaultProps.level;
        this.prefix = config.components.prefix;
    }
}
FormMessageComponent.defaultProps = {
    className: '',
    level: FORM_MESSAGE_LEVELS.INFO
};
FormMessageComponent.decorators = [
    { type: Component, args: [{
                selector: `${config.components.prefix}-form-message }`,
                template: "<div [ngClass]=\"[prefix + '-form-message', className, level]\">\n  <ng-content></ng-content>\n</div>\n"
            }] }
];
FormMessageComponent.propDecorators = {
    className: [{ type: Input, args: ['className',] }],
    level: [{ type: Input, args: ['type',] }]
};
if (false) {
    /** @type {?} */
    FormMessageComponent.defaultProps;
    /** @type {?} */
    FormMessageComponent.prototype.className;
    /** @type {?} */
    FormMessageComponent.prototype.level;
    /** @type {?} */
    FormMessageComponent.prototype.prefix;
}
