// materialize-angular/esm5/app/completed-components/drawer/drawer.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/drawer/drawer.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
var DrawerModel = /** @class */ (function () {
    function DrawerModel() {
    }
    return DrawerModel;
}());
export { DrawerModel };
if (false) {
    /** @type {?} */
    DrawerModel.prototype.className;
    /** @type {?} */
    DrawerModel.prototype.closeOnBackdrop;
    /** @type {?} */
    DrawerModel.prototype.hasBackdrop;
    /** @type {?} */
    DrawerModel.prototype.isOpen;
    /** @type {?} */
    DrawerModel.prototype.position;
    /** @type {?} */
    DrawerModel.prototype.transitionDuration;
    /** @type {?} */
    DrawerModel.prototype.type;
}
/** @enum {string} */
var DRAWER_POSITIONS = {
    RIGHT: "right",
    LEFT: "left",
};
export { DRAWER_POSITIONS };
/** @enum {string} */
var DRAWER_TYPES = {
    OVER: "over",
    PUSH: "push",
    STATIC: "static",
};
export { DRAWER_TYPES };
