// materialize-angular/esm2015/app/utils/get-supported-events.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/utils/get-supported-events.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
/** @type {?} */
const MOUSE_EVENTS = {
    down: 'mousedown',
    move: 'mousemove',
    resize: 'resize',
    up: 'mouseup'
};
/** @type {?} */
const TOUCH_EVENTS = {
    down: 'touchstart',
    move: 'touchmove',
    resize: 'resize',
    up: 'touchend'
};
/**
 * @return {?}
 */
export function supportTouchEvents() {
    return 'ontouchstart' in window;
}
/** @type {?} */
export const supportedEvents = (/**
 * @return {?}
 */
() => {
    /** @type {?} */
    const events = supportTouchEvents()
        ? TOUCH_EVENTS
        : MOUSE_EVENTS;
    return events;
});
