// materialize-angular/esm2015/app/completed-components/slider/slider-option/slider-option.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/slider/slider-option/slider-option.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { config } from '../../../config';
export class SliderOptionComponent {
    constructor() {
        this.className = SliderOptionComponent.defaultProps.className;
        this.value = SliderOptionComponent.defaultProps.value;
        this.prefix = config.components.prefix;
        this.isActive = false;
        this.onClickEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    onClick() {
        this.onClickEmitter.emit(this.value);
    }
}
SliderOptionComponent.defaultProps = {
    className: '',
    value: null
};
SliderOptionComponent.decorators = [
    { type: Component, args: [{
                selector: `${config.components.prefix}-slider-option }`,
                template: "<div [ngClass]=\"[prefix + '-slider-option-container', className]\"\n  [class.active]=\"isActive\"\n\n  (click)=\"onClick()\"\n  #template>\n  <ng-content></ng-content>\n</div>\n"
            }] }
];
/** @nocollapse */
SliderOptionComponent.ctorParameters = () => [];
SliderOptionComponent.propDecorators = {
    templateRef: [{ type: ViewChild, args: ['template', { static: true },] }],
    className: [{ type: Input }],
    value: [{ type: Input }],
    onClickEmitter: [{ type: Output, args: ['onClick',] }]
};
if (false) {
    /** @type {?} */
    SliderOptionComponent.defaultProps;
    /** @type {?} */
    SliderOptionComponent.prototype.templateRef;
    /** @type {?} */
    SliderOptionComponent.prototype.className;
    /** @type {?} */
    SliderOptionComponent.prototype.value;
    /** @type {?} */
    SliderOptionComponent.prototype.onClickEmitter;
    /** @type {?} */
    SliderOptionComponent.prototype.prefix;
    /** @type {?} */
    SliderOptionComponent.prototype.isActive;
}
