// materialize-angular/esm2015/app/completed-components/glossary/glossary-item/glossary-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/glossary/glossary-item/glossary-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, EventEmitter, HostListener, Input, Output } from '@angular/core';
import { config } from '../../../config';
export class GlossaryItemComponent {
    constructor() {
        this.className = GlossaryItemComponent.defaultProps.className;
        this.referenceId = GlossaryItemComponent.defaultProps.referenceId;
        this.prefix = config.components.prefix;
        this.isActive = false;
        this.onClickEmitter = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        this.onClickEmitter.emit(this.referenceId);
    }
}
GlossaryItemComponent.defaultProps = {
    className: '',
    referenceId: ''
};
GlossaryItemComponent.decorators = [
    { type: Component, args: [{
                selector: `${config.components.prefix}-glossary-item`,
                template: "<li\n  [ngClass]=\"[prefix + '-glossary-item', className]\"\n  [class.active]=\"isActive\"\n>\n  <ng-content></ng-content>\n</li>\n"
            }] }
];
/** @nocollapse */
GlossaryItemComponent.ctorParameters = () => [];
GlossaryItemComponent.propDecorators = {
    onClickEmitter: [{ type: Output, args: ['onClick',] }],
    className: [{ type: Input }],
    referenceId: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    GlossaryItemComponent.defaultProps;
    /** @type {?} */
    GlossaryItemComponent.prototype.onClickEmitter;
    /** @type {?} */
    GlossaryItemComponent.prototype.className;
    /** @type {?} */
    GlossaryItemComponent.prototype.referenceId;
    /** @type {?} */
    GlossaryItemComponent.prototype.prefix;
    /** @type {?} */
    GlossaryItemComponent.prototype.isActive;
}
