// materialize-angular/esm2015/app/completed-components/scroll-spy/scroll-spy-item/scroll-spy-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/scroll-spy/scroll-spy-item/scroll-spy-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
import { Component, ElementRef, Input } from '@angular/core';
import { config } from '../../../config';
export class ScrollSpyItemComponent {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
        this.className = ScrollSpyItemComponent.defaultProps.className;
        this.id = ScrollSpyItemComponent.defaultProps.id;
    }
}
ScrollSpyItemComponent.defaultProps = {
    className: '',
    id: ''
};
ScrollSpyItemComponent.decorators = [
    { type: Component, args: [{
                selector: `${config.components.prefix}-scroll-spy-item`,
                template: "<div [attr.id]=\"id\">\n  <ng-content></ng-content>\n</div>\n"
            }] }
];
/** @nocollapse */
ScrollSpyItemComponent.ctorParameters = () => [
    { type: ElementRef }
];
ScrollSpyItemComponent.propDecorators = {
    className: [{ type: Input }],
    id: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ScrollSpyItemComponent.defaultProps;
    /** @type {?} */
    ScrollSpyItemComponent.prototype.className;
    /** @type {?} */
    ScrollSpyItemComponent.prototype.id;
    /** @type {?} */
    ScrollSpyItemComponent.prototype.element;
}
