// materialize-angular/esm2015/app/completed-components/calendar/calendar.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: app/completed-components/calendar/calendar.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Workylab. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://raw.githubusercontent.com/workylab/materialize-angular/master/LICENSE
 */
/**
 * @record
 */
export function CalendarModel() { }
if (false) {
    /** @type {?} */
    CalendarModel.prototype.className;
    /** @type {?} */
    CalendarModel.prototype.date;
    /** @type {?} */
    CalendarModel.prototype.displayOtherMonthDays;
}
/**
 * @record
 */
export function DateModel() { }
if (false) {
    /** @type {?} */
    DateModel.prototype.date;
    /** @type {?} */
    DateModel.prototype.dayLabel;
    /** @type {?} */
    DateModel.prototype.ISODate;
    /** @type {?} */
    DateModel.prototype.isToday;
    /** @type {?} */
    DateModel.prototype.isOutOfMonth;
    /** @type {?} */
    DateModel.prototype.monthLabel;
    /** @type {?} */
    DateModel.prototype.showSelected;
}
/**
 * @record
 */
export function MonthModel() { }
if (false) {
    /** @type {?} */
    MonthModel.prototype.label;
    /** @type {?} */
    MonthModel.prototype.number;
    /** @type {?} */
    MonthModel.prototype.year;
}
/**
 * @record
 */
export function DateLabel() { }
if (false) {
    /** @type {?} */
    DateLabel.prototype.name;
    /** @type {?} */
    DateLabel.prototype.shortName;
    /** @type {?} */
    DateLabel.prototype.shortestName;
}
/**
 * @record
 */
export function DayLabels() { }
if (false) {
    /** @type {?} */
    DayLabels.prototype.sunday;
    /** @type {?} */
    DayLabels.prototype.monday;
    /** @type {?} */
    DayLabels.prototype.tuesday;
    /** @type {?} */
    DayLabels.prototype.wednesday;
    /** @type {?} */
    DayLabels.prototype.thursday;
    /** @type {?} */
    DayLabels.prototype.friday;
    /** @type {?} */
    DayLabels.prototype.saturday;
}
/**
 * @record
 */
export function MonthLabels() { }
if (false) {
    /** @type {?} */
    MonthLabels.prototype.january;
    /** @type {?} */
    MonthLabels.prototype.february;
    /** @type {?} */
    MonthLabels.prototype.march;
    /** @type {?} */
    MonthLabels.prototype.april;
    /** @type {?} */
    MonthLabels.prototype.may;
    /** @type {?} */
    MonthLabels.prototype.june;
    /** @type {?} */
    MonthLabels.prototype.july;
    /** @type {?} */
    MonthLabels.prototype.august;
    /** @type {?} */
    MonthLabels.prototype.september;
    /** @type {?} */
    MonthLabels.prototype.october;
    /** @type {?} */
    MonthLabels.prototype.november;
    /** @type {?} */
    MonthLabels.prototype.december;
}
